### PR TITLE
memorizer: initial build setup

### PR DESCRIPTION
**Pull request by ndauten**
- What I did

Create initial Memorizer project subdirectory.

- How I did it

Followed default kernel build process. There are a few things left out like perf build to get it building.

- How to verify it

Standard project build, push, moby build, and linuxkit test.

- Description for the changelog

- A picture of a cute animal (not mandatory but encouraged)

### CODE REVIEW

**Comment by ndauten** [comment]
The current status is the system builds but does not have a clean tree. So it doesn't have a docker image pushed to the hub yet. I wanted to get this PR setup so that I could start interacting and ask a few questions.

The build works fine, but when running the minimal .yml it appears to run out of max stack size and gets stuck. This same configuration builds and boots fine outside of the linuxkit environment.

@riyazdf a few questions:

license: my initial license on the memorizer src was NSCA/Illinois, I presume I need to change this?
Is there any special kernel config options that I need to be aware of when setting things up?
Is there a way that I can rebuild in the same docker image? It has to do a full compile every time.
Is there a way to get a shell in the image the Dockerfile is creating so that I can manually do a few things?
Once you let me know about the source I can add the diff patch. I want to make sure I do things as appropriate.

Overall though, I think the boot is hanging. It occurs right after the LinuxKit welcome message and whale are printed.

**Comment by justincormack** [comment]
@ndauten try removing `console=tty0` and moving the `getty` into the `init` section (see https://github.com/linuxkit/linuxkit/tree/master/pkg/getty#linuxkit-debug) it may give better debug output if something is failing. Not 100% sure the stack message is the issue - looks like you just have more kernel debug options enabled than the default setup.

**Comment by ndauten** [comment]
@justincormack looks like containerd is booting but it has trouble finding mount points? Is this something due to a configuration option? 

```
INFO[0000] starting containerd boot...                   module=containerd
INFO[0000] starting debug API...                         debug="/run/containerd/debug.sock" module=containerd
INFO[0000] loading plugin "io.containerd.content.v1.content"...  module=containerd type=io.containerd.content.v1
INFO[0000] loading plugin "io.containerd.snapshotter.v1.btrfs"...  module=containerd type=io.containerd.snapshotter.v1
INFO[0000] loading plugin "io.containerd.snapshotter.v1.overlayfs"...  module=containerd type=io.containerd.snapshotter.v1
INFO[0000] loading plugin "io.containerd.differ.v1.base-diff"...  module=containerd type=io.containerd.differ.v1
INFO[0000] loading plugin "io.containerd.metadata.v1.bolt"...  module=containerd type=io.containerd.metadata.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.containers"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.content"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.diff"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.monitor.v1.cgroups"...  module=containerd type=io.containerd.monitor.v1
INFO[0000] loading plugin "io.containerd.runtime.v1.linux"...  module=containerd type=io.containerd.runtime.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.tasks"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.healthcheck"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.images"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.namespaces"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.snapshots"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.version"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] starting GRPC API server...                   module=containerd
INFO[0000] containerd successfully booted in 0.030108s   module=containerd
EOF
container_linux.go:262: starting container process caused "process_linux.go:261: applying cgroup configuration for process caused \"mountpoint for devices not found\""
 - 000-dhcpcd
Starting service: "getty"
INFO[0000] new shim started                              module="containerd/execution" socket="/var/lib/containerd/io.containerd.runtime.v1.linux/default/getty/shim.sock"
FATA[0000] failed to create task                         error="rpc error: code = Unknown desc = runtime create failed: runc create failed: container_linux.go:262: starting container process caused \"process_linux.go:261: applying cgroup configuration for process caused \\\"mountpoint for devices not found\\\"\"" service=getty
```

**Comment by ndauten** [comment]
@justincormack see above for more recent output on the problem. Sorry for the closed PR problem.

**Comment by justincormack** [comment]
We have a kernel config test container which checks what containers need.
(Sorry, out of office at the minute)

**Comment by ndauten** [comment]
I removed the dhcp on boot, but it still gave me an error about mount points: 

```
Starting containerd

Welcome to LinuxKit

                        ##         .
                  ## ## ##        ==
               ## ## ## ## ##    ===
           /"""""""""""""""""\___/ ===
          {                       /  ===-
           \______ O           __/
             \    \         __/
              \____\_______/

INFO[0000] starting containerd boot...                   module=containerd
INFO[0000] starting debug API...                         debug="/run/containerd/debug.sock" module=containerd
INFO[0000] loading plugin "io.containerd.content.v1.content"...  module=containerd type=io.containerd.content.v1
INFO[0000] loading plugin "io.containerd.snapshotter.v1.btrfs"...  module=containerd type=io.containerd.snapshotter.v1
INFO[0000] loading plugin "io.containerd.snapshotter.v1.overlayfs"...  module=containerd type=io.containerd.snapshotter.v1
INFO[0000] loading plugin "io.containerd.differ.v1.base-diff"...  module=containerd type=io.containerd.differ.v1
INFO[0000] loading plugin "io.containerd.metadata.v1.bolt"...  module=containerd type=io.containerd.metadata.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.containers"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.content"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.diff"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.monitor.v1.cgroups"...  module=containerd type=io.containerd.monitor.v1
INFO[0000] loading plugin "io.containerd.runtime.v1.linux"...  module=containerd type=io.containerd.runtime.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.tasks"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.healthcheck"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.images"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.namespaces"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.snapshots"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.version"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] starting GRPC API server...                   module=containerd
INFO[0000] containerd successfully booted in 0.028983s   module=containerd
Starting service: "getty"
INFO[0000] new shim started                              module="containerd/execution" socket="/var/lib/containerd/io.containerd.runtime.v1.linux/default/getty/shim.sock"
FATA[0000] failed to create task                         error="rpc error: code = Unknown desc = runtime create failed: runc create failed: container_linux.go:262: starting container process caused \"process_linux.go:261: applying cgroup configuration for process caused \\\"mountpoint for devices not found\\\"\"" service=getty
```

**Comment by riyazdf** [comment]
@justincormack: are you referring to https://github.com/linuxkit/linuxkit/blob/master/test/pkg/kernel-config/check-kernel-config.sh ?

@ndauten: do you have the same issue without getty running?  I'm trying to decipher the error message...

@ndauten:
> license: my initial license on the memorizer src was NSCA/Illinois, I presume I need to change this?

I'm not too familiar with the NSCA/Illinois license.  We can probably have a separate license for this subcomponent of the project but I'm not sure what the best path here is.

> Is there a way that I can rebuild in the same docker image? It has to do a full compile every time.

I'm not sure I understand your question here.  Are you asking if you can rebuild parts of the kernel image?  Or is this for output of a `moby build`?

> Is there a way to get a shell in the image the Dockerfile is creating so that I can manually do a few things?

for the kernel image?  Not sure if there's a good way of doing this

**Comment by ndauten** [comment]
@riyazdf @justincormack apparently the issue was resolved with one of the configuration options I copied from the 4.9 kernel config. So this is great progress! I am getting very noisy output from one of my debug tools (kasan) so it's hard to fully see what's going on, but I drop into the getty namespace shell on boot. Thanks for the help. I will continue working on getting the default setup configured properly, and then once that's done I will push an image to the linuxkit docker hub, so that others can start using right away. I've got some designs ready for optimizations to make it more useful. 

I will create an issue to discuss what other think is a good workflow for processing memorizer data.

**Comment by ndauten** [comment]
@riyazdf 

> License

I can change to whatever. I suppose  we are dealing with 2 sets: LinuxKit/Docker and Linux. I just used that as a default, it is basically an MIT. 

> I'm not sure I understand your question here. Are you asking if you can rebuild parts of the kernel image? Or is this for output of a moby build?

Basically for building the kernel image. Each time it does a complete new build. I'm wondering if we can persist the build state within the image used by the Dockerfile, otherwise, the test-debug cycle using LinuxKit is a bit slower.

Of course I can always do this outside, but having the builder is very nice for dev process.

> > Is there a way to get a shell in the image the Dockerfile is creating so that I can manually do a few things?
> for the kernel image? Not sure if there's a good way of doing this
For the kernel image. So basically get a shell prompt. Of course I could pull an image (e.g., the alpine one) and build inside by hand, but was wondering if that is the only option.

**Comment by rn** [comment]
> Basically for building the kernel image. Each time it does a complete new build. I'm wondering if we can persist the build state within the image used by the Dockerfile, otherwise, the test-debug cycle using LinuxKit is a bit slower.

Not easily and we don't really want that in the build system. I just use a fast machine for kernel compiles. However for kernel dev you can also just compile on your linux machine and create a image manually. the format in the container image is pretty simple

**Comment by rn** [comment]
this needs a rebase and DCO sign

**Comment by ndauten** [comment]
When doing the `rebase -i` I cannot see the particular commit lacking a signature. It was a merge commit. Any ideas?

**Comment by rn** [comment]
?? don't merge. just be on your branch and to a `git rebase master` after you pulled upstream

**Comment by ndauten** [comment]
Yeah, I have two machines where I've got this checked out, and accidentally entered a merge. I wonder if there is anyway to unmerge, or get the visibility to squash the merge message?

**Comment by justincormack** [comment]
You can create a new branch from master and then cherry pick all the commits in and then reset this branch to that one (or open a new PR)

**Comment by ndauten** [comment]
@justincormack i modified the log here for that merge to be signed, but if you'd prefer I can create a new branch, cherry pick, and reset the branch. Sorry for my git learning.

**Comment by ndauten** [comment]
Status Update:

I have most of the system and bugs all worked out! There are a few rough edges that persist:

1) KASAN is tripping on out-of-bounds stack accesses from KSAN itself, which shouldn't happen, but causes a lot of log messages.

2) Licensing. I am tracing down my official rules for release. So until that comes back source will be held back. To alleviate this I think I can push the binary image to linuxkitprojects hub and minimally that permits usage. I can then send code to interested folks as needed. 

Also we have some enhancements currently being made to the tool by @jpandey2 to optimize the runtime.

**Comment by ndauten** [comment]
@riyazdf Finished initial memorizer addition with some minimal documentation. The KASAN output is annoying and should be addressed in a subsequent commit.

File: projects/memorizer/docs/memorizer.txt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+# clear free objects: the current system traces from boot with a lot of
+# uninteresting data
+echo 1 > /sys/kernel/debug/clear_object_list
+# enable memorizer object access tracking, which by default is off
+echo 1 > /sys/kernel/debug/memorizer_log_access
+# Now run whatever test
+tar zcf something.tar.gz /somedir &
+ssh u@h:/somefile 
+...
+# Disable access loggin
```
Comment (riyazdf) [inline projects/memorizer/docs/memorizer.txt]: nit typo: `logging`

**Comment by riyazdf** [comment]
@ndauten: in addition to squashing, it'd be nice to add memorizer to this list on [`projects/README.md`](https://github.com/linuxkit/linuxkit/blob/master/projects/README.md#current-projects)

**Comment by ndauten** [comment]
@riyazdf I added line to projects/README.md and squashed the commits.

**Pull request merged**